csharp/src/Drivers/Apache/Spark/CloudFetch/SparkCloudFetchReader.cs

```diff
@@ -0,0 +1,269 @@
+        /// <param name="isLz4Compressed">Whether the results are LZ4 compressed.</param>
+        public SparkCloudFetchReader(HiveServer2Statement statement, Schema schema, bool isLz4Compressed)
+        {
+            this.statement = statement;
+            this.schema = schema;
+            this.isLz4Compressed = isLz4Compressed;
+        }
+
+        /// <summary>
+        /// Gets the Arrow schema.
+        /// </summary>
+        public Schema Schema { get { return schema; } }
+
+        private HttpClient HttpClient
+        {
+            get { return _httpClient.Value; }
+        }
+
+        /// <summary>
+        /// Reads the next record batch from the result set.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The next record batch, or null if there are no more batches.</returns>
+        public async ValueTask<RecordBatch?> ReadNextRecordBatchAsync(CancellationToken cancellationToken = default)
+        {
+            while (true)
+            {
+                // If we have a current reader, try to read the next batch
+                if (this.currentReader != null)
+                {
+                    RecordBatch? next = await this.currentReader.ReadNextRecordBatchAsync(cancellationToken);
+                    if (next != null)
+                    {
+                        return next;
+                    }
+                    else
+                    {
+                        this.currentReader.Dispose();
+                        this.currentReader = null;
+                    }
+                }
+
+                // If we have more links to process, download and process the next one
+                if (this.resultLinks != null && this.linkIndex < this.resultLinks.Count)
+                {
+                    var link = this.resultLinks[this.linkIndex++];
+                    byte[]? fileData = null;
+
+                    // Retry logic for downloading files
+                    for (int retry = 0; retry < MaxRetries; retry++)
+                    {
+                        try
+                        {
+                            fileData = await DownloadFileAsync(link.FileLink, cancellationToken);
+                            break; // Success, exit retry loop
+                        }
+                        catch (Exception ex) when (retry < MaxRetries - 1)
+                        {
+                            // Log the error and retry
+                            Console.WriteLine($"Error downloading file (attempt {retry + 1}/{MaxRetries}): {ex.Message}");
+                            await Task.Delay(RetryDelayMs * (retry + 1), cancellationToken);
+                        }
+                    }
+
+                    if (fileData == null)
+                    {
+                        // All retries failed, continue to the next link or fetch more links
+                        continue;
```

Comment:
I'm a little concerned about the failure mode here and in line 150. Won't this mean that the user silently gets less data and doesn't even know about it?